Policy fix mapstate owners

jrajahalme opened the pull request:
When an identity is deleted, the agent processes the incremental policy
map changes against previously saved policy state with owners and
dependents of the identity key to decide if an identity can be marked for
deletion. In some cases, the `owners` field for the identity entry was
nil (due to them being redacted when returned by ConsumeMapChanges()
previously), and the deleteKeyWithChanges would return early without
tracking the identity for deletion. As a result, when this information is
consumed later to delete the identity from the BPF maps, the identity was
being leaked in those cases.

Callers of ConsumeMapChanges() and AddVisibilityKeys() depend on the
returned added and deleted keys to perform datapath policy map updates
without iterating whole policy maps. Moreover, the "realized" version of
endpoint's policy map used to be updated solely based on these returned
entries.

Make this less fragile and error prone by only returning the added and
deleted map keys (no values). AddVisibilityKeys() supports reverting the
changes, so it still needs to return the full MapState (keys and values)
for any changed entries.

Now that ConsumeMapChanges() and AddVisibilityKeys() return the keys of
added entries, the caller needs to find the value to be inserted from the
desired policy map using the given key.

Fixes: #18581
Fixes: 0d585f110846237eb3611e88d0992ba26a162f72
Fixes: https://github.com/cilium/cilium/issues/19003
Co-authored-by: Aditi Ghag <aditi@cilium.io>
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

```release-note
Fixed a bug where deleted identities would remain in BPF policy maps.
```

jrajahalme committed:
policy: Only return keys for incremental updates

When an identity is deleted, the agent processes the incremental policy
map changes against previously saved policy state with owners and
dependents of the identity key to decide if an identity can be marked for
deletion. In some cases, the `owners` field for the identity entry was
nil (due to them being redacted when returned by ConsumeMapChanges()
previously), and the deleteKeyWithChanges would return early without
tracking the identity for deletion. As a result, when this information is
consumed later to delete the identity from the BPF maps, the identity was
being leaked in those cases.

Callers of ConsumeMapChanges() and AddVisibilityKeys() depend on the
returned added and deleted keys to perform datapath policy map updates
without iterating whole policy maps. Moreover, the "realized" version of
endpoint's policy map used to be updated solely based on these returned
entries.

Make this less fragile and error prone by only returning the added and
deleted map keys (no values). AddVisibilityKeys() supports reverting the
changes, so it still needs to return the full MapState (keys and values)
for any changed entries.

Now that ConsumeMapChanges() and AddVisibilityKeys() return the keys of
added entries, the caller needs to find the value to be inserted from the
desired policy map using the given key.

Fixes: #18581
Fixes: 0d585f110846237eb3611e88d0992ba26a162f72
Co-authored-by: Aditi Ghag <aditi@cilium.io>
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

aditighag commented on pkg/endpoint/bpf.go:
```diff
@@ -1242,13 +1231,25 @@ func (e *Endpoint) applyPolicyMapChanges() (proxyChanges bool, err error) {
-		if !e.addPolicyKey(keyToAdd, entry, true) {
+		if e.addBPFPolicyKey(keyToAdd, entry, true) {
+			// Operation was successful, add to realized state.
+			e.realizedPolicy.PolicyMapState[keyToAdd] = entry
```
>so that the update can be made to the correct map in
syncDesiredPolicyMapWith().

Not sure I understand the motivation for this move. Both the places are pointing to the same policy state map, no?

jrajahalme replied:
Here yes, but on the other call site not always (map is passed down as a parameter).

joestringer replied:
:+1: The `syncPolicyMapWithDump()` case passes not the endpoint's stored realized state map, but instead it takes a fresh dump of the "real" realized state and uses that for comparison (not this line of code, lower down). LGTM.

aditighag commented on pkg/endpoint/bpf.go:
```diff
@@ -1087,7 +1087,7 @@ func (e *Endpoint) updatePolicyMapPressureMetric() {
 // the deleted entry had a proxy port assigned to it.  *hadProxy is
 // not otherwise changed (e.g., it is never set to 'false').
-func (e *Endpoint) deletePolicyKey(keyToDelete policy.Key, incremental bool, hadProxy *bool) bool {
+func (e *Endpoint) deleteBPFPolicyKey(keyToDelete policy.Key, incremental bool) bool {
```
Nit : The function description needs to be updated.

jrajahalme replied:
Thanks, fixed.

twpayne reviewed (commented):
It looks like the tests in `pkg/policy/mapstate_test.go` need to be updated as they still rely on the `export()` method.

aditighag commented on pkg/endpoint/bpf.go:
```diff
@@ -1110,26 +1108,17 @@ func (e *Endpoint) deletePolicyKey(keyToDelete policy.Key, incremental bool, had
 }
 
-func (e *Endpoint) addPolicyKey(keyToAdd policy.Key, entry policy.MapStateEntry, incremental bool) bool {
+func (e *Endpoint) addBPFPolicyKey(keyToAdd policy.Key, entry policy.MapStateEntry, incremental bool) bool {
```
`addBPFPolicyKey` and `deleteBPFPolicyKey` aren't symmetrical anymore. Is that intentional? Can we get `entry` at the caller site too?

jrajahalme replied:
These functions should still be symmetrical, the realized policy map manipulation was removed from both.

aditighag commented on pkg/endpoint/bpf.go:
```diff
@@ -1110,26 +1108,17 @@ func (e *Endpoint) deletePolicyKey(keyToDelete policy.Key, incremental bool, had
-		"incremental":         incremental,
-	}, "deletePolicyKey")
+		logfields.BPFMapKey: keyToDelete,
+		"incremental":       incremental,
```
The log messaging printing both key and entry with owners and dependents was useful during debugging. Since we are retrieving entry at the consumer site, can we add a new log there?

jrajahalme replied:
I'll add the entry the parameters of `deleteBPFPolicyKey()` so that we can keep logging it.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -1110,26 +1108,17 @@ func (e *Endpoint) deletePolicyKey(keyToDelete policy.Key, incremental bool, had
-
-	// Operation was successful, remove from realized state.
-	delete(e.realizedPolicy.PolicyMapState, keyToDelete)
 	e.updatePolicyMapPressureMetric()
```
The `e.updatePolicyMapPressureMetric()` call is intended to update the metrics for policy map usage whenever the contents change. I think this call should also move to the caller to ensure that it is correctly updated whenever the map is modified.

Ditto for `addBPFPolicyKey()`.

jrajahalme replied:
BPF policy map is still updated in this function (now called `deleteBPFPolicyKey()`), the caller updates the endpoint's realized policy map. Maybe the intent of the metric is to indicate the pressure on the BPF maps instead of the Go maps? If that is the case then the metric update is in the right place :-)

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -459,9 +460,9 @@ var visibilityDerivedFrom = labels.LabelArrayList{visibilityDerivedFromLabels}
-func (keys MapState) AddVisibilityKeys(e PolicyOwner, redirectPort uint16, visMeta *VisibilityMetadata, adds, deletes MapState) {
+// for both added and changed keys. 'deletes' contains the old values for changed keys.
+// This function does not delete any keys.
+func (keys MapState) AddVisibilityKeys(e PolicyOwner, redirectPort uint16, visMeta *VisibilityMetadata, adds Keys, oldValues MapState) {
```
This code changes the named return value for `deletes`, please update the comment to reflect this change.

jrajahalme replied:
Thanks, done!

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -1202,23 +1202,31 @@ func (e *Endpoint) applyPolicyMapChanges() (proxyChanges bool, err error) {
 			if redirectPort, exists := e.realizedRedirects[proxyID]; exists && redirectPort != 0 {
-				e.desiredPolicy.PolicyMapState.AddVisibilityKeys(e, redirectPort, visMeta, adds, deletes)
+				e.desiredPolicy.PolicyMapState.AddVisibilityKeys(e, redirectPort, visMeta, adds, nil)
 			}
 		}
 		for _, visMeta := range e.visibilityPolicy.Egress {
 			proxyID := policy.ProxyID(e.ID, visMeta.Ingress, visMeta.Proto.String(), visMeta.Port)
 			if redirectPort, exists := e.realizedRedirects[proxyID]; exists && redirectPort != 0 {
-				e.desiredPolicy.PolicyMapState.AddVisibilityKeys(e, redirectPort, visMeta, adds, deletes)
+				e.desiredPolicy.PolicyMapState.AddVisibilityKeys(e, redirectPort, visMeta, adds, nil)
```
Why do we no longer care about the `oldValues` here?

jrajahalme replied:
`AddVisibilityKeys()` does not delete any keys so I changed the semantics of it to not return `deleted` values, but `oldValues`. The old values of any changed (or deleted) keys are only needed if the changes can be reverted, and that is not the case here.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -1202,23 +1202,31 @@ func (e *Endpoint) applyPolicyMapChanges() (proxyChanges bool, err error) {
 
+		entry, exists := e.desiredPolicy.PolicyMapState[keyToAdd]
+		if !exists {
+			e.getLogger().WithFields(logrus.Fields{
+				logfields.AddedPolicyID: keyToAdd,
+			}).Warn("Tried adding policy map key not in policy")
+			continue
```
Is this an expected condition, reliant on a race condition? Or is it an unexpected condition that would demonstrate a bug? I'm mostly wondering whether there's any possibility that, for instance, in a high churn cluster with identities being introduced & removed all the time, could this warning message be triggered regularly? Could it flood the logs? Is the plan for us to gather any instances of this warning and use them to investigate further issues?

aditighag replied:
The main motivation of changing `MapState` type to `Keys` is to avoid hitting this condition as we don't allow callers to erroneously mutate owners of a `Key`. `consumeMapChanges` (ideally) should be the only one doing book-keeping of owners and hence removal of keys from the policy map.
So this condition hopefully allows us to identify the cases where this doesn't hold true during the add operation. (I didn't exactly answer your question about the end goal, just giving more context. :))
@jrajahalme Thinking out loud - do we need a reciprocal condition in the delete path? A key is being deleted but it exists in the policy map?

jrajahalme replied:
This should be an internal (to the agent) error condition. I'll add unit tests covering addition and deletion of an entry on the same round of incremental updates, even though the code implementing it is pretty clearly removing the key from adds or deletes, if the key us deleted or added, respectively.

joestringer reviewed (changes requested):
I must admit, I did not fully follow what's going on here. Would probably take another one or two attempts to get my head around what's really going on. I picked out a couple of small nits, most notably that the metrics would not be synced correctly in the current state.

joestringer reviewed (approved):
I follow now, the regular policy generation path would correctly generate the entries with an 'owner', but the incremental policy path through `ConsumeMapChanges()` would deliberately drop the owner references there. The latter path would then trigger an unintentional override of the `desiredPolicy` map's entries to replace the entries. The replaced entries would no longer contain the owner references.

Then `addKeyWithChanges()` / `deleteKeyWithChanges()` don't really have any handling for the case where the owner is nil / unknown. So once those entries with nil owners end up in the `desiredPolicy` map, they get stuck there and Cilium would always resolve the intended datapath state from that datapath state. Hence the leak.

jrajahalme committed:
endpoint: Rename syncDesiredPolicyMapWith() as syncPolicyMapsWith()

syncDesiredPolicyMapWith() syncs the diffs between the endpoint't desired
policy mapstate and the given 'realized' mapstate to the datapath bpf
map, applying all differences also to endoint's realized mapstate. The
current name of this function highlights the desired mapstate, which
might be understood as endpoint's realized mapstate not being
updated. Clarify this by renaming syncDesiredPolicyMapWith() as
syncPolicyMapsWith().

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -1283,14 +1283,16 @@ func (e *Endpoint) syncPolicyMap() error {
 // difference between the given 'realized' and desired policy state without
 // dumping the bpf policy map.
-func (e *Endpoint) syncDesiredPolicyMapWith(realized policy.MapState, withDiffs bool) (diffCount int, diffs []policy.MapChange, err error) {
+// Changes are synced to endpoint's realized policy mapstate, 'realized' is
+// not modified.
```
I'm not sure I quite follow the meaning in this last sentence.

It looks to me like we go through the desired policy map, and apply all desired changes that the daemon thinks have not yet been realized. So it's more that the _bpf_ policymaps are synced to the _desired_ state based on the daemon's understanding of the realized state.

joestringer replied:
The whole terminology here of "sync desired policy map with ( realized )" just seems kinda confusing to me. The goal here is to sync the BPF policy maps with the desired state, right? But we just make it more efficient by using the current realized state view.

joestringer replied:
(thinking aloud if there's another less confusing naming convention here..)

joestringer replied:
I think that part of my confusion above is that "sync `<x>` with `<y>`" can be interpreted in a way that describes the way that the information is flowing, basically that `<x>` is the thing that changes and `<y>` is the source of truth. But right now, the thing that actually changes (bpf) is not explicitly mentioned, and it's implied that the desired policy changes here... which doesn't seem to be the case.

joestringer replied:
Looking a little further, with the latest version of this PR, this statement seems to be incorrect: `'realized' is not modified.`

`addPolicyKey()` and `deletePolicyKey()` are still modifying the `realizedPolicy`.

jrajahalme replied:
This function can be called with a temporary 'realized' map, dumped from the datapath, or with the endpoint's realized policy map. Either way diffs between this passed in map and endpoint's desired policy map are applied to the bpf policy map *and* the endpoint's realized policy map, not the passed in 'realized' map. For the call where 'realized' is the endpoint's realized policy map this does not mean much, but in the case where 'realized' is a temporary map dumped from the datapath there is no point updating it after computing the diff (as the temporary map is thrown away right after), and this way the endpoint's own realized policy map becomes (more) synced with the datapath bpf map.

joestringer replied:
Ah, okay - I was wondering why this 'realized' map was explicitly handed in, as I assumed that this `realized` map was exactly the same as `e.realizedPolicy`, but that's not the case. This stuff is pretty subtle :thinking:
